WIP: kappa-next

Frando opened the pull request:
This PR is the current state of my efforts to make kappa-core work with different kinds of sources.
It is based on the #13 and incoroprates a lot of useful feedback from noffle, substack and others.

* Makes kappa-core dependency-less
* Includes breaking changes, thus the `kappa-classic` export retaining api compatibility with current kappa-core is not possible anymore.

Breaking changes are:

* the `use` function takes a name, a **source**, and a view
* both sources and views can expose APIs, they are mounted on `kappa.view[name]` and `kappa.source[name]`
* sources are expected to keep track of state, so that each call to `pull` only ever returns messages which haven't been returned before

I added documentation for the new API, see the [rendered README](https://github.com/Frando/kappa-core/blob/kappa5-new/README.md).

I think this is ready for review now. If wanted, I can also do a writeup of the changes? But maybe its quite obvious already also from the new README and tests.

Frando committed:
Move index.js to kappa-old.js

Frando committed:
Add new Kappa!

This is a dependency-free version of the kappa core idea. It connects
sources to views and tracks state.

Next commit will add a kappaClassic wrapper that support multifeed by default.

Frando committed:
Add kappaClassic wrapper and source handlers

Frando committed:
Add thunky dependency

Frando committed:
Fixes

Frando committed:
Add example as test

Frando committed:
Cleanup

Frando committed:
Docs

Frando committed:
Rename method

Frando committed:
Add flowchart

Frando committed:
Minor changes to wiring

Frando committed:
Add a hyperdrive source handler & test

Frando committed:
Enable all tests

Frando committed:
Use svg for flow graph

Frando committed:
Added kappa-graph.svg

Frando committed:
Use svg for flow graph

Frando committed:
Merge branch 'kappa5' of github.com:Frando/kappa-core into kappa5

Frando committed:
Remove png graph

Frando committed:
Default opts, and cleanup

Frando committed:
Remove subflows for now

Frando committed:
Fix hyperdrive, improve tests

Frando committed:
Tests & cleanup

Frando committed:
Lint standard

Frando committed:
Make methods public

Frando committed:
Cleanup open & ready

Also removes a "parent flow" leftover of the removed subflows pattern.

Frando committed:
Better multifeed backward compatibility

Frando committed:
docs

Frando committed:
context for views

ameba23 committed:
bump corestore to 4.0.0

ameba23 committed:
add corestore source

ameba23 committed:
add test for corestore source

ameba23 committed:
add default error handler for hypercore source

ameba23 committed:
valueencoding = json in corestore test

ameba23 committed:
function for corestore constructor

Frando committed:
Add transform opt

Frando committed:
Add stacked views.

Frando committed:
clearIndex for stacked views

Frando committed:
Merge remote-tracking branch 'cobox/corestoreSource' into kappa5

Frando committed:
Rework API after discussions and feedback.

A kappa now is a list of flows where each connects a single source to a
single view.

Hyperdrive source and stacked views still needs to be updated. Other tests pass.

Frando committed:
Finish API rework. Ready for review.

Frando committed:
version

Frando committed:
Use level-mem for state helper

Frando committed:
Cleanup

Frando committed:
Improve source utils

Frando committed:
Add feed api on multifeed and corestore sources

Frando committed:
fix

Frando committed:
fix

Frando committed:
Streamline and remove unneeded opts

Frando committed:
docs

Frando committed:
docs

Frando committed:
docs

Frando committed:
docs

Frando commented:
Here's an example on what modules using kappa-core would need to change: https://ledger-git.dyne.org/CoBox/kappa-drive/pulls/36/files

I also added a commit that leaves `kappa.api[name]` available as an alias to `kappa.view[name]`.

Frando committed:
fix: storeversion only after open

Frando committed:
Use pipeline helper

Frando committed:
add close logic

Frando committed:
if source has ready method await it in flow.ready

Frando committed:
fix onsourceready

m4gpi commented:
hey @noffle, cobox plans on supporting @Frando's upgrade to kappa-core, as it provides some pretty important flexibility that we're going to need when managing our indexes. I'd be interested to hear your thoughts on this PR! :)

hackergrrl commented:
@Frando is this PR ready for review? I saw WIP in the title so I assumed you're still working on it.

Frando committed:
fix open logic

Frando committed:
support ready on source directly

Frando committed:
support opts.filter for flows

Frando committed:
missing dev deps

Frando committed:
filter empty messages

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+* `name` (string) the name of the flow, has to be unique per kappa core
+* `source` object with properties:
+  * `open: function (flow, next)`: (optional) Handler to call on open
+  * `pull: function (next)`: Pull new messages from the view. Should call `next` with either nothing or an object that looks like this:
```
I have a preference toward instead calling `next` like

`next([messages], finished, cb)`

This is partly my preference (easier to type), but also means skipping an object allocation **per source pull**. This is a hot path so it'd be great to minimize object allocs.

(I also have a preference for making `finished` default to true.)

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+* `name` (string) the name of the flow, has to be unique per kappa core
+* `source` object with properties:
+  * `open: function (flow, next)`: (optional) Handler to call on open
+  * `pull: function (next)`: Pull new messages from the view. Should call `next` with either nothing or an object that looks like this:
```
How does clean-up work? Do we need a `close` function?

Frando replied:
Yep, there's a close function. Added docs.

hackergrrl commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 node_modules
 package-lock.json
+pnpm-lock.yaml
```
(comment unrelated to this file)

Hey do you think we ought to update `example.js` to show how the new API works? That or nix it and just have a good README example.

hackergrrl commented on README.md:
````diff
@@ -1,5 +1,102 @@
+    ```
+    The source has to track its state, so that subsequent calls to `pull()` do not return the same messages. Use the `onindexed` callback to update state.
+
+    There are several source handlers included in kappa-core (TODO: document sources). See the tests and sources directories.
````
TODO: document sources :D

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+
+    There are several source handlers included in kappa-core (TODO: document sources). See the tests and sources directories.
+
+    A simple state handler that perists state in a leveldb (or in memory) is included and used by the bundled source handler. It's available for use by custom sources on `Kappa.SimpleState`. If the bundled source handlers get a `db` option passed with levelup instance, the source state will be persisted.
```
`SimpleState` is not currently being exported.

Frando replied:
Fixed.

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+
+    A simple state handler that perists state in a leveldb (or in memory) is included and used by the bundled source handler. It's available for use by custom sources on `Kappa.SimpleState`. If the bundled source handlers get a `db` option passed with levelup instance, the source state will be persisted.
+
+  * `reset: function (cb)`: Reset internal state (next pull should start at the beginning). This handler is called by the kappa-core if the view's `version` property increases and thus wants to restart indexing.
```
What are these bullet pointed functions describing?

Frando replied:
That's still the source API. Moved things around to make it clear.

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+
+Both `source` and `view` can have an `api` property with an object of function. The functions are exposed on `kappa.view[name]` / `kappa.source[name]`. Their `this` object refers to the flow, and their first parameter is the `kappa`. Other parameters are passed through.
+
+#### `kappa.reset(name, cb)`
```
What's the downstream usecase for this? I think if nobody we know needs it, we could make it hidden, and keep the API surface area down, which means less maintenance for us!

Frando replied:
I think it can be made hidden, yes. Instead advertize in the API to increase the view version number if you want to trigger a reindex. I used it in testing and benchmarking situations (eg to measure how long indexing takes).

hackergrrl replied:
Sounds good. I agree with keeping it around undocumented for tests & benchmarks.

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+
+#### `kappa.ready(names, cb)`
+
+Call `cb` after all flows with a name in the `names` array have finished processing. If `names` is empty, all flows will be awaited.
```
```suggestion
Call `cb` exactly once, after all flows with a name in the `names` array have finished processing. If `names` is empty, all flows will be awaited. If all the flows are already ready, `cb` is called immediately.
```

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+      for (let [key, value] of Object.entries(source.api)) {
+        this.source[key] = bindFn(value, this, this.context)
+      }
+      delete source.api
```
Why delete these view & source keys from outside arguments?

Frando replied:
No particular reason. Let's keep them.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    if (this.opts.filter) this._transform.push(this.opts.filter)
+    if (this.opts.transform) this._transform.push(this.opts.transform)
+    if (this._view.filter) this._transform.push(this._view.filter.bind(this._view))
+    if (this._view.transform) this._transform.push(this._view.transform.bind(this._view))
```
What are the use cases for all of these transform and filter hooks? I'm -1 on having them if we don't have a use yet, so we can keep complexity down.

Frando replied:
The use caes for the transform hook is that at several times, I wanted to have a slightly different version of a source that emits messages in a different format to match what a view expects. With `opts.transform` this can be done for a specific flow, `view.transform` and `map.transform` allows to extend existing views/sources and adapt them. `filter` was just the same so I removed this. I think the little added complexity by the transform step is worth the gained flexibility in composition. I'll rethink this another time soon when I go over my existing uses once more to reflect on the APIs.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+  }
+
+  get version () {
+    return this._view.version || 1
```
This will return `1` if version is `0`, which is maybe what the user wanted. If we check for a "good value" (non-negative integer?) when it's passed in at `kappa.use`, we could avoid having this logic at all.

Frando replied:
Sounds good, changed this

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    function onopen () {
+      if (--pending !== 0) return
+      if (self._source.fetchVersion) {
+        self._source.fetchVersion((err, version) => {
```
We could emit the error here if the source is unable to fetch its version.

Frando replied:
sound good, done

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+      } else ondone()
+    }
+
+    function ondone () {
```
As per above comment, we could funnel errors into here & emit them.

Frando replied:
fixed this too

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    }
+  }
+
+  ready (cb, waitForSource) {
```
`waitForSource` isn't used.

Frando replied:
fixed

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    this.pause()
+    let pending = 1
+    process.nextTick(() => {
+      if (this._view.clearIndex) ++pending && this._view.clearIndex(done)
```
Could you document `clearIndex` in the README?

hackergrrl replied:
I know it'd break backwards compat (which we're mostly doing anyways!), but having `clearIndex` be `reset` would make it the same as the source API, which'd be nice.

Frando replied:
SGTM. Changed to `reset`, still allowing `clearIndex` also for now, can be removed before release though likely.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    let pending = 1
+    process.nextTick(() => {
+      if (this._view.clearIndex) ++pending && this._view.clearIndex(done)
+      if (this._source.reset) ++pending && this._source.reset(done)
```
Is `reset` required on the source? It probably should be. Could you mark the various fields for `source` in the README as optional/required?

Frando replied:
Yup I think it should be required. Currently this is not enforced, but this would mean that after increasing the view's version number, if the source does not implement `reset`, things go wrong (the view would continue receiving new messages instead of starting from the beginning). I think there might be usecases for "non-historic" sources, but that has untested implications. I marked it in the documentation as required. Likely we want to assert this in the code as well?

hackergrrl replied:
What do you think about checking sources/flows/views on `kappa.use()` to ensure they have all of the properties set, and fail if not? I bet that will catch a lot of mistakes.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    }
+  }
+
+  update () {
```
Could you document this API in the README?

Frando replied:
Done

hackergrrl commented on README.md:
```diff
@@ -1,5 +1,102 @@
+This source exposes an API method `feed (key)` that return a feed by key from the underlying corestore.
+
+---
+
```
(reminder for us to revise the old README, which is in tact below the above additions)

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    this._source.pull(onbatch)
+
+    function onbatch (result) {
+      // console.log(self.name, 'onbatch', result)
```
```suggestion
```

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+
+      const { messages = [], finished, onindexed } = result
+
+      // TODO: Handle timeout / error?
```
Right now in Mapeo and probably Cabal we return errors (like writes to storage) via `view.map`'s `next` callback. It would be helpful if that continued to be handled & propagated up the stack.

Frando replied:
Yup errors are passed and emitted on the kappa, fixed the docs.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    function close (err, result) {
+      if (err) self.emit('error', err)
```
```suggestion
      if (err) return self.emit('error', err)
```

Frando replied:
Fixed this. However one thing I'm not sure about: If a view errors, should the flow be stopped / paused / run again? Currently, I think it would run in a loop which is not intended. Likely we want a new Error state for flows.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+  get api () { return this.view }
+
+  use (name, source, view, opts = {}) {
+    opts.status = opts.status || this.status
```
What is the use case for passing a non-ready state to a flow?

Frando replied:
I think the usecase was to have a flow start in Paused state.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    if (typeof names === 'string') names = [names]
+    if (!names) names = Object.keys(this.flows)
+
+    let pending = names.length
```
If there are no views, `cb` will never get called.

Frando replied:
Fixed this.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    let flows = Object.values(this.flows)
+    let pending = flows.length
+    if (!pending) done()
+    // TODO: Propagate errors?
```
Yeah, I think catching the first error and passing it up to `cb` (once) would be good here.

Frando replied:
Fixed this.

hackergrrl commented on package.json:
```diff
@@ -1,8 +1,8 @@
-  "author": "Stephen Whitmore <sww@eight.net>",
-  "version": "4.0.1",
+  "author": "Stephen Whitmore <sww@eight.net>, Franz Heinzmann <frando@unbiskant.org>",
+  "version": "7.0.0-alpha1",
```
Can we leave the version change out of the commit & do an `npm version major` after this is merged into master?

Frando replied:
SGTM

hackergrrl commented on sources/hypercore.js:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  reset (cb) {
+    this.state.set(0, cb)
```
There is another `reset` below. I'm guessing this one is dead code since it refers to `this,state.set` which doesn't exist.

hackergrrl commented on sources/hypercore.js:
```diff
@@ -0,0 +1,76 @@
+
+  pull (cb) {
+    this.state.get(this.name, (err, seq) => {
+      if (err) return cb(err)
```
TODO: handle getting an error back from `pull`'s `next` callback

Frando replied:
Yes this depends on what the arguments to the callback for `pull` will look like. Leaving for now, as that's one of the API change questions I think.

hackergrrl commented on sources/hypercore.js:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    feed.getBatch(at, to, { wait: false }, (err, res) => {
+      if (err) return next()
```
I think I've left a few scattered comments to this effect, but I think it'd be worthwhile to have `next`'s signature be `next(err, messages, finished, onindexed)`.

hackergrrl commented on sources/multifeed.js:
```diff
@@ -0,0 +1,44 @@
+    }
+  }
+
+  function onfeed (flow, feed, cb) {
```
`cb` not used

Frando replied:
Fixed

hackergrrl commented on sources/util/merge.js:
```diff
@@ -0,0 +1,41 @@
+exports.mergePull = function (sources, next) {
+  if (!sources.length) return next()
```
Zalgo!
```suggestion
  if (!sources.length) return process.nextTick(next)
```

Frando replied:
fixed

hackergrrl commented on sources/util/merge.js:
```diff
@@ -0,0 +1,41 @@
+    let messages = []
+    let finished = true
+    for (let result of results) {
+      if (result.messages) messages = messages.concat(result.messages)
```
You can skip an array allocation on this hot path with this trick:
```suggestion
      if (result.messages) Array.prototype.push.apply(messages, result.messages)
```

Frando replied:
Nice, yes!

hackergrrl commented on sources/multifeed.js:
```diff
@@ -0,0 +1,44 @@
+const hypercoreSource = require('./hypercore')
```
I :heart: that you rewrote `multifeed-index` in 44 lines!

hackergrrl commented on sources/util/state.js:
```diff
@@ -0,0 +1,78 @@
+
+function noop () {}
+
+class FakeDB {
```
I was thinking it might be good to have the callbacks in this class use `process.nextTick`, since other implementations will likely have an async API.

Frando replied:
Makes sense

hackergrrl reviewed (commented):
Hey @Frando :wave:  Awesome PR, I really like the conceptual & API designs of this, and am eager to iterate on this with you & get it merged. I've reviewed everything but the tests, which I'll try and look at early next week.

Frando commented:
Wow thanks for this review @noffle, cool! I'll try to get to this soon, but might take me a bit more than a week as I'm quite busy with a deadline next week.

noffle commented on test/basic.js:
```diff
@@ -1,36 +1,155 @@
+  kappa.source.view1.push(1)
+  kappa.source.view1.push(2)
+  kappa.source.view2.push(3)
+  kappa.source.view2.push(4)
```
This is a bigger convo, but the API format of

`kappa.source.flowName.METHOD`

might feel nicer as

`kappa.flowName.source.METHOD`

What do you think? This is kinda a big change at this point, but it'll be an even bigger pain to make later on once even more downstreamers are using this API.

hackergrrl commented on kappa.js:
```diff
@@ -0,0 +1,319 @@
+    else close()
+    function close () {
+      let pending = 1
+      if (self._source.close) ++pending && self._source.close(cb)
+      if (self._view.close) ++pending && self._view.close(cb)
```
```suggestion
      if (self._source.close) ++pending && self._source.close(done)
      if (self._view.close) ++pending && self._view.close(done)
```

hackergrrl replied:
Almost missed this one :D

Frando replied:
Ha :D fixed

hackergrrl commented on test/corestore.js:
```diff
@@ -0,0 +1,46 @@
+    const core2 = store.get({ valueEncoding: 'json' })
+    const db = mem()
+
+    kappa.use('view', corestoreSource({ store, db }), createSimpleView())
```
I :heart: how easy it is to write views on top of a totally different source!

hackergrrl commented on test/multifeed.js:
```diff
@@ -0,0 +1,71 @@
+  return sumview
+}
+
+// function replicate (a, b, opts, cb) {
```
unused

hackergrrl commented on test/stacked-view.js:
```diff
@@ -0,0 +1,152 @@
+const ram = require('random-access-memory')
+const hypercoreSource = require('../sources/hypercore')
+
+// Has to be updated.
```
TODO, or perhaps remove for now?

jwerle commented:
very exciting to watch this!

Frando commented:
So finally I'll get back here this week, sorry that it took so long. 

I think I'll go over @noffle's review in two steps: 

* First, I'll try to handle everything non-API-breaking (onto the current state of things) which I'll push onto this PR
* Then, I'll compile an overview of the proposed API changes, possibly adding some that I myself am thinking about so that we can then discuss these once more before committing to them. 

As far as I know Sonar and Cobox are currently depending on this branch, if there are others please chime in too!

Frando committed:
open in nextTick

Frando committed:
first round of fixes after review

Frando committed:
more fixes and doc improvements after review

Frando committed:
remove filter option for sources and views

Frando committed:
optimization: ensure good version value, dont remove api props

Frando committed:
support errors on open

Frando committed:
rename view.clearIndex to view.reset

Frando committed:
docs formatting

Frando committed:
Docs improvements

Frando committed:
more fixes from nettle's review

Frando committed:
remove outdated test

Frando committed:
corrections

Frando commented:
I finished a big round of non-api breaking changes from @noffle's review. Yay! I added a couple of comments to things I did not address right away.

Now, to the proposed API changes:

* The main proposed API change by @noffle is to change the signature of the `next` callback in a source's `pull` handler from `next({ messages, finished, onindexed })` to `next(err, messages, finished, onindexed)`. And, at the same time, change the default for `finished` to true. I think I like both changes. 

* I have another change I'd propose that is purely asthetic: I'd maybe like to rename `ready` to `sync`. It seems clearer to me, as a kappa is never "ready" if it's listening for remote sources that can update at any time. Also it may be confusing as oftenly in our ecosystem `ready` is an `open` alias.

* One thing I've been thinking about: Currently kappa-core injects itself as a first argument to the API functions. I added a `context` op to make this overridable with another object, as oftenly I've seen me needing the "object in which the kappa is used" (e.g. a cabal, a kappa-record-db) and not the kappa-core itself. At the same time, this can also be passed as an option into the `createView` or `createSource` function. So I thought if we should just remove this injected first argument to the API functions. Thoughts?

*  I have an improved version for the `SimpleState` utility that I wrote for `kappa-sparse-indexer`. I'll add this too once I start with the breaking changes.

* The internal state for each flow currently is either Ready, Running, Paused. Likely we want to have an Error state also, into which a flow is put when either a view returned an error from `map` or a source returned an error from `pull`

* Maybe we want to expose a `getState` function that returns the current state of things, and also asks the source for the indexing state (as in the recently added feature to `multifeed-index`) 

I am using this API in `kappa-sparse-indexer` and `kappa-record-db`, which I'd update accordingly then. I think cobox uses only sources internal to kappa-core. @kyphae or did you implement custom sources somewhere? Also if someone else apart from me and @kyphae is using this branch in dependencies, could you give a quick ping here? I'd like to know how "free" I am at the moment with regard to API changes. I've seen @substack using it in examples at least. For those currently depending on it - you can pin to `github:Frando/kappa-core#exp-0.2.3` to not get API breaking changes right away if I push to this branch.

ghost commented:
I was already expecting some changes so pushing is fine for me.

As far as the argument-prepending goes, I find that I very rarely use this feature because I will almost always do `kcore.use(someView(arg1, arg2, ...))` where `someView` is `module.exports = function (arg1, arg2...) {}` so where I need the kcore instance or other arguments I pass it in there. Or more commonly, I will pass in particular exported apis instead of the whole kappa-core instance.

Frando committed:
Prepare for API changes

Frando committed:
add error state

Frando committed:
track indexing state

* allow sources to return a state object that describes
  the indexing progress
* exposes a getState method
* emit a state-update event

Frando committed:
mark status property private

Frando committed:
move test util libs

Frando committed:
cleanup

Frando committed:
remove unneeded deps

Frando committed:
fix corestore source

Frando committed:
Emit state-update event on kappa

Frando committed:
Reuse iteration code on kappa

Frando committed:
Better state tracking

Frando committed:
Expose top-level getState

Frando committed:
Restructure test SimpleSource

hackergrrl commented:
:wave:! Hi @Frando, sorry it's taken so long to get back to you on this. It seems like every time there are updates I need to sit down for a long time and download it all into my brain again. :)  I appreciate the summary of proposed changes.

> The main proposed API change by @noffle is to change the signature of the next callback in a source's pull handler from next({ messages, finished, onindexed }) to next(err, messages, finished, onindexed). And, at the same time, change the default for finished to true. I think I like both changes.

Great!

> I have another change I'd propose that is purely asthetic: I'd maybe like to rename ready to sync. It seems clearer to me, as a kappa is never "ready" if it's listening for remote sources that can update at any time. Also it may be confusing as oftenly in our ecosystem ready is an open alias.

Do you think it might be confusing, since it's so similar in meaning to `replicate`? Could `indexesReady()` work, or something else that hints at this being a means to wait on the views to catch up? If you disagree, feel free to use `sync` -- this doesn't feel worth bikeshedding too much on.

> One thing I've been thinking about: Currently kappa-core injects itself as a first argument to the API functions. I added a context op to make this overridable with another object, as oftenly I've seen me needing the "object in which the kappa is used" (e.g. a cabal, a kappa-record-db) and not the kappa-core itself. At the same time, this can also be passed as an option into the createView or createSource function. So I thought if we should just remove this injected first argument to the API functions. Thoughts?

I'm down with this change. I *really* regret the kludge I chose of injecting fake arguments.

> The internal state for each flow currently is either Ready, Running, Paused. Likely we want to have an Error state also, into which a flow is put when either a view returned an error from map or a source returned an error from pull

Yes please.

> Maybe we want to expose a getState function that returns the current state of things, and also asks the source for the indexing state (as in the recently added feature to multifeed-index)

I'm ok with punting on this until after these changes get merged. This PR is also very big, so anything we can do to keep the scope down and change later would be :fire:.
